segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadOperation.java

```diff
@@ -141,23 +143,56 @@ private void logEnd() {
                     } else {
                         Preconditions.checkState(bytesToRead != 0, "bytesToRead is 0");
                         // Read data from the chunk.
-                        return chunkedSegmentStorage.getChunkStorage().openRead(chunkToReadFrom.getName())
-                                .thenComposeAsync(chunkHandle ->
-                                        chunkedSegmentStorage.getChunkStorage().read(chunkHandle,
-                                                currentOffset.get() - startOffsetForCurrentChunk.get(),
-                                                bytesToRead,
-                                                buffer,
-                                                currentBufferOffset.get())
-                                                .thenAcceptAsync(bytesRead -> {
-                                                    bytesRemaining.addAndGet(-bytesRead);
-                                                    currentOffset.addAndGet(bytesRead);
-                                                    currentBufferOffset.addAndGet(bytesRead);
-                                                    totalBytesRead.addAndGet(bytesRead);
-                                                }, chunkedSegmentStorage.getExecutor()),
-                                        chunkedSegmentStorage.getExecutor());
+                        return CompletableFuture.runAsync(() -> {
+                            // Create parallel requests to read each chunk.
+                            chunkReadFutures.add(readChunk(chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get()));
+                            log.trace("{} read - reading chunk - op={}, segment={}, chunk={} offset={} length={} bufferOffset={}",
+                                    chunkedSegmentStorage.getLogPrefix(),
+                                    System.identityHashCode(this), handle.getSegmentName(),
+                                    chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get());
+                            bytesRemaining.addAndGet(-bytesToRead);
+                            currentOffset.addAndGet(bytesToRead);
```

Comment:
Also, even if one of the parallel reads fail, the counters would have been changed anyway. Are you testing this scenario?

Reply:
The counters are calculating what chunks to read and at what offset in byte array to copy contents. 
This calculation does not change whether actual reads are parallel or sequential.

Reply:
readData returns Futures.allOf(chunkReadFutures)